protocol/src/test/scala/io/mediachain/protocol/DataObjectGenerators.scala

```diff
@@ -33,49 +39,87 @@ object DataObjectGenerators {
 
-  val genArtefactChainCell = for {
-    artefact <- genReference
-    chain <- genReference
+
+  def genEntityChainCell(
+    entityGen: Gen[Entity] = genEntity,
+    chainGen: Gen[Option[Reference]] = genOptionalReference
+  ) = for {
+    entity <- genReferenceFor(entityGen)
+    chain <- chainGen
+    meta <- genMeta
+  } yield EntityChainCell(entity, chain, meta)
+
+  def genArtefactChainCell(
+    artefactGen: Gen[Artefact] = genArtefact,
+    chainGen: Gen[Option[Reference]] = genOptionalReference
+  ) = for {
+    artefact <- genReferenceFor(artefactGen)
+    chain <- chainGen
     meta <- genMeta
-  } yield ArtefactChainCell(artefact, Some(chain), meta)
+  } yield ArtefactChainCell(artefact, chain, meta)
 
-  val genEntityUpdateCell = for {
-    base <- genEntityChainCell
+  def genEntityUpdateCell(
+    entityGen: Gen[Entity] = genEntity,
+    chainGen: Gen[Option[Reference]] = genOptionalReference
+  ) = for {
+    base <- genEntityChainCell(entityGen, chainGen)
   } yield EntityUpdateCell(base.entity, base.chain, base.meta)
 
-  val genEntityLinkCell = for {
-    base <- genEntityChainCell
-    entityLink <- genReference
+  def genEntityLinkCell(
+    entityGen: Gen[Entity] = genEntity,
+    chainGen: Gen[Option[Reference]] = genOptionalReference,
+    entityLinkGen: Gen[Reference] = genReference
+  ) = for {
+    base <- genEntityChainCell(entityGen, chainGen)
+    entityLink <- entityLinkGen
   } yield EntityLinkCell(base.entity, base.chain, base.meta, entityLink)
 
-  val genArtefactUpdateCell = for {
-    base <- genArtefactChainCell
+  def genArtefactUpdateCell(
+    artefactGen: Gen[Artefact] = genArtefact,
+    chainGen: Gen[Option[Reference]] = genOptionalReference
+  ) = for {
+    base <- genArtefactChainCell(artefactGen, chainGen)
   } yield ArtefactUpdateCell(base.artefact, base.chain, base.meta)
 
-  val genArtefactCreationCell = for {
-    base <- genArtefactChainCell
-    entity <- genReference
+  def genArtefactCreationCell(
+    artefactGen: Gen[Artefact] = genArtefact,
+    chainGen: Gen[Option[Reference]] = genOptionalReference,
+    entityGen: Gen[Reference] = genReference
+  ) = for {
+    base <- genArtefactChainCell(artefactGen, chainGen)
+    entity <- entityGen
   } yield ArtefactCreationCell(base.artefact, base.chain, base.meta, entity)
 
-  val genArtefactDerivationCell = for {
-    base <- genArtefactChainCell
-    artefactOrigin <- genReference
+  def genArtefactDerivationCell(
+    artefactGen: Gen[Artefact] = genArtefact,
+    chainGen: Gen[Option[Reference]] = genOptionalReference,
+    artefactOriginGen: Gen[Reference] = genReference
+  ) = for {
+    base <- genArtefactChainCell(artefactGen, chainGen)
+    artefactOrigin <- artefactOriginGen
   } yield ArtefactDerivationCell(base.artefact, base.chain, base.meta, artefactOrigin)
 
-  val genArtefactOwnershipCell = for {
-    base <- genArtefactChainCell
-    entity <- genReference
+  def genArtefactOwnershipCell(
```

Comment:
I'm not sure we need these (i.e. specialized types beyond Update), though that may be a separate discussion. We're getting kind of close to the point where we need to have a serious think about how all this stuff is modeled.